src/provider/model/Model.ts

```diff
@@ -131,73 +167,57 @@ export class Model {
         return this.findWindows(appWindow => appWindow.appInfo.name === name);
     }
 
-    private async onWindowCreated(identity: Identity, manifestUrl: string): Promise<void> {
-        // Registration is asynchronous and sensitive to race conditions. We use a deferred promise
-        // to signal to other sensitive operations that it is safe to proceed.
-        const deferredPromise = new DeferredPromise();
-        const pendingRegistration = deferredPromise.promise;
-        const resolvePending = deferredPromise.resolve;
-
-        this._pendingRegistrations.set(getId(identity), pendingRegistration);
+    private onWindowSeen(identity: Identity): void {
+        this.getOrCreateExpectedWindow(identity, 'seen');
+    }
 
+    private async onWindowCreated(identity: Identity, manifestUrl: string): Promise<void> {
         const apps = await this._directory.getAllApps();
         const appInfoFromDirectory = apps.find(app => app.manifest.startsWith(manifestUrl));
 
-        const id: string = getId(identity);
-        // If the app is not in directory we ignore it. We'll add it to the model if and when it connects to FDC3
-        if (appInfoFromDirectory && !this._windowsById[id]) {
+        if (appInfoFromDirectory) {
+            // If the app is in directory, we register it immediately
             this.registerWindow(appInfoFromDirectory, identity, true);
-        } else if (this._windowsById[id]) {
-            console.info(`Ignoring window created event for ${id} - window was already registered`);
-        }
+        } else {
+            // If the app is not in directory, we'll add it to the model if and when it connects to FDC3
+            allowReject(this.getOrCreateExpectedWindow(identity).connected.then(async () => {
```

Comment:
The lack of `await` here is intentional, but I don't *think* it'd make any difference in practice to have an await. I'm saying "in response to this event, set this thing up to happen", rather than "in response to this event, wait, possibly a very long time, for this to happen"

`connected` will reject if the window is close, which makes sense, since we know it will never connect, and is consistent with the behaviour needed here to have `expectWindow` behave as expected

Note `allowReject` is around the promise returned by `then` (not `connected` itself), which I believe will itself reject if `connected` rejects